### ISSUE TITLE
Simplify not working with matrix multiplication?

### ISSUE THREAD

**Issue by filonik** [josdejong/mathjs#2614]
Shouldn't the following produce the same result?

```
import * as math from 'mathjs'

let scope = {
  A: math.matrix([
    [1, 2],
    [3, 4],
  ]),
  v: math.matrix([5, 6]),
}

const r0 = math.evaluate('A * v', scope)
console.log('Evaluate', r0) // Evaluate: DenseMatrix {_data: [17, 39]}

//const A = new math.SymbolNode('A')
//const v = new math.SymbolNode('v')
//const expr = new math.OperatorNode('*', 'multiply', [A, v])
const expr = math.parse('A * v')

const r1 = math.simplify(expr, scope)
console.log('Simplify', r1) // Simplify: ConstantNode {value: 0}
```

When I run the code, `evaluate` produces the expected output (`[17, 39]`), but `simplify` does not (instead it produces `0` which doesn't seem right).

**Comment by josdejong** [comment]
Thanks for reporting, that sounds like a bug, `0` is indeed a wrong result for that expression.

Anyone able to help debugging/fixing this issue in `simplify`? Help would be welcome.

**Comment by brunoSnoww** [comment]
After digging up a little bit the issue seems to be on [line 205 of simplifyCore](https://github.com/josdejong/mathjs/blob/61a21da6b7df7c2eb262889250e378f7d44a958a/src/function/algebra/simplifyCore.js#L205) . The isZero(a0.value) returns a DenseMatrix with all entries equal to zero, and inside the 'if' this is evaluated to true, and therefore a node0 ( a node of value 0 ) is returned. But even if we refactor the isZero function and make this code return an OperatorNode at [line 223](https://github.com/josdejong/mathjs/blob/61a21da6b7df7c2eb262889250e378f7d44a958a/src/function/algebra/simplifyCore.js#L223), the following stack trace appears: 

```
Error
    at new DimensionError (node_modules/mathjs/lib/esm/error/DimensionError.js:18:16)
    at matAlgo13xDD (fnode_modules/mathjs/lib/esm/type/matrix/utils/matAlgo13xDD.js:37:13)
    at DenseMatrix, DenseMatrix (node_modules/mathjs/lib/esm/type/matrix/utils/matrixAlgorithmSuite.js:43:47)
    at generic (node_modules/typed-function/lib/umd/typed-function.js:1782:29)
    at equal (node_modules/typed-function/lib/umd/typed-function.js:1817:24)
    at _exactMatch (node_modules/mathjs/lib/esm/function/algebra/simplify.js:1223:12)
    at mergeMatch (node_modules/mathjs/lib/esm/function/algebra/simplify.js:962:16)
    at combineChildMatches (node_modules/mathjs/lib/esm/function/algebra/simplify.js:988:18)
    at Array.reduce (<anonymous>)
    at mergeChildMatches (node_modules/mathjs/lib/esm/function/algebra/simplify.js:1006:29) {
  actual: 2,
  expected: 1,
  relation: undefined,
  message: 'Dimension mismatch (2 != 1)'
```

Which I guess is because the matAlgo13xDD.js expects matrices of equal sizes, and in our example they do not have.

So basically, I have two questions:

- Shouldn't the isZero function just return a single boolean ? ( which in be the case of matrices and arrays will be all entries equal to zero ? )

- Some suggestion of where to look at.  Maybe some refactor on this  matAlgo13xDD.js could solve the problem ? Or make such handling of matrices inside simplifyCore itself 

Thanks ! :)

**Comment by soleilcot** [comment]
I took a look at this as it appears the issue still occurs. The reason for `0` being returned lies in the `simplifyCore.js` logic that executes as part of the simplification rule set:

https://github.com/josdejong/mathjs/blob/4a26b068d8803cfde6495552d1c8dc410e0cbe07/src/function/algebra/simplifyCore.js#L205-L223 

In this logic, both the matrix nodes are considered `ConstantNode` and `isZero` is also truthy for the `DenseMatrix` node values. The reason `isZero` is truthy in this case is because the return for a matrix argument to `isZero` is itself a matrix of booleans resulting from an element wise execution of `isZero`. In this context, `simplifyCore` simplifies the whole expression to `0`.

Skipping that one condition wouldn't be enough though. It seems that most the simplification rules are for use in a one-dimensional algebraic context in which there can be unknowns and are thus not needed for constant matrix simplification. Instead it seems to me that we would need a new, separate set of logic specifically for matrix simplification and evaluation.

**Comment by josdejong** [comment]
Thanks for debugging this @soleilcot , this makes sense indeed.

So as a first step we may need to add a check on whether the value we're operating on is a numeric value and not a Matrix or something else I think? Like adding a check `isNumeric(a0).value`?
